File: euler40.py
# Найдите значение следующего выражения числа Чамбертоуна:
# d1 × d10 × d100 × d1000 × d10000 × d100000 × d1000000

from datetime import datetime
from functools import reduce


def euler40() -> None:
    """
    Решение задачи Эйлера №40.

    Найдите значение произведения цифр числа Чамбертоуна.
    """
    n = 1_000_000

    # 1 вариант
    start_time = datetime.now()
    champernowne = ''.join(str(_) for _ in range(n + 1))
    positions = [10 ** _ for _ in range(len(str(n)))]
    result = reduce(
        lambda x, y: x * y,
        [int(champernowne[pos]) for pos in positions]
    )
    print(result)
    print(datetime.now() - start_time)

    # 2 вариант
    start_time = datetime.now()
    pass
    print(datetime.now() - start_time)


if __name__ == "__main__":
    euler40()

File: go/euler40.go
// euler40 находит произведение цифр числа Чамбертоуна.
//
// Решение задачи Эйлера №40.
package main

import (
	"fmt"
	"time"
)

func euler40() {
	n := 1000000

	startTime := time.Now()
	champernowne := ""
	for i := 0; i <= n; i++ {
		champernowne += fmt.Sprintf("%d", i)
	}
	positions := make([]int, 0)
	for i := 0; i < len(fmt.Sprintf("%d", n)); i++ {
		positions = append(positions, powInt(10, i))
	}
	result := 1
	for _, pos := range positions {
		if pos < len(champernowne) {
			digit := int(champernowne[pos] - '0')
			result *= digit
		}
	}
	fmt.Println(result)
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	fmt.Println(time.Since(startTime))
}

func powInt(base, exp int) int {
	result := 1
	for i := 0; i < exp; i++ {
		result *= base
	}
	return result
}

func main() {
	euler40()
}
